### PR TITLE
feat(appsync): expose the AppSyncDomain of the custom domain of an AppSync api

### PR DESCRIPTION
**Pull request by AKoetsier**
When configuring a custom domain on an AppSync api you couldn't easily get the AppSyncDomainName attribute of the associated AWS::AppSync::DomainName. This fix will expose that attribute so that the provided example doesn't throw errors.

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md)

### Adding new Unconventional Dependencies:

* [ ] This PR adds new unconventional dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md/#adding-new-unconventional-dependencies)

### New Features

* [ ] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md)?
	* [ ] Did you use `yarn integ` to deploy the infrastructure and generate the snapshot (i.e. `yarn integ` without `--dry-run`)?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Review by mrgrain — DISMISSED** [review]
Thanks for the contribution!

Let's make the name a bit more user friendly please.

And out of curiosity, what is the difference between `appSyncDomainName` and the provided domain (`myDomainName` in the example)?

**Comment by mrgrain** [comment]
Additionally, please make sure the build is passing. You can see the latest build failure here: https://github.com/aws/aws-cdk/pull/21554#issuecomment-1211930951

tl;dr Your change to the example is not compiling.

File: packages/@aws-cdk/aws-appsync/lib/graphqlapi.ts [hunk 1]
```diff
@@ -470,6 +470,13 @@ export class GraphqlApi extends GraphqlApiBase {
    */
   public readonly apiKey?: string;
 
+  /**
+   * the associated custom AppSync domain, if present
+   *
+   * @default - no custom associated domain
+   */
+  public readonly domainNameAppSyncDomainName?: string;
```
[the remark this answers is not in the document]
Comment (AKoetsier) [reply packages/@aws-cdk/aws-appsync/lib/graphqlapi.ts:478]: Ok I will make this change. I was trying to follow the naming used in the apigateway domain package.

**Comment by AKoetsier** [comment]
> And out of curiosity, what is the difference between `appSyncDomainName` and the provided domain (`myDomainName` in the example)?

The myDomainName in the example is just the string 'api.example.com'. So the example would create a CNAME with the name api.example.com with a value api.example.com. The appSyncDomainName is the cloudfront url of the newly associated custom domain.

File: packages/@aws-cdk/aws-appsync/lib/graphqlapi.ts [hunk 2]
```diff
@@ -470,6 +470,13 @@ export class GraphqlApi extends GraphqlApiBase {
    */
   public readonly apiKey?: string;
 
+  /**
+   * the associated custom AppSync domain, if present
+   *
+   * @default - no custom associated domain
+   */
+  public readonly appSyncDomainName?: string;
+
```
Comment (mrgrain) [inline packages/@aws-cdk/aws-appsync/lib/graphqlapi.ts:479]:
```suggestion
  /**
   * The associated domain name provided by AppSync
   */
  public readonly appSyncDomainName: string;

```
`domainName.attrAppSyncDomainName` seems to be always available. Is there a reason why you made this optional at first?

File: packages/@aws-cdk/aws-appsync/README.md [hunk 3]
```diff
@@ -330,7 +330,7 @@ const zone = route53.HostedZone.fromHostedZoneAttributes(this, `HostedZone`, {
 new route53.CnameRecord(this, `CnameApiRecord`, {
   recordName: 'api',
   zone,
-  domainName: myDomainName,
+  domainName: api.appSyncDomainName!,
```
Comment (mrgrain) [inline packages/@aws-cdk/aws-appsync/README.md:333]:
```suggestion
  domainName: api.appSyncDomainName,
```

**Review by mrgrain — CHANGES_REQUESTED** [review]

File: packages/@aws-cdk/aws-appsync/lib/graphqlapi.ts (hunk 2 above)
Comment (AKoetsier) [reply packages/@aws-cdk/aws-appsync/lib/graphqlapi.ts:479]: The domainName resource is only created when you pass the domainName properties in the api. So when creating a minimal api it will not create an AWS::AppSync::DomainName or an AWS::AppSync::DomainNameApiAssociation and therefore you don't have the appSyncDomainName.

**Comment by AKoetsier** [comment]
I converted the attribute to a getter function so that we can throw an error when no domainName is configured. This is more inline with other modules and makes it easier to use.

File: packages/@aws-cdk/aws-appsync/lib/graphqlapi.ts (hunk 2 above)
Comment (AKoetsier) [reply packages/@aws-cdk/aws-appsync/lib/graphqlapi.ts]: Sorry I only just saw this comment appear. I have changed it a bit to be even more clear when you are requesting the appSyncDomainName without a domainName configuration.

**Comment by mrgrain** [comment]
> I converted the attribute to a getter function so that we can throw an error when no domainName is configured. This is more inline with other modules and makes it easier to use.

Thanks! I was about to comment about that. 😄

**Review by mrgrain — APPROVED** [review]
Looking good now. Ready to ship if the build passes. 🎉
